### PR TITLE
Best-checkpoint weight soup: average top-3 run checkpoints

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

**Best-checkpoint EMA-soup: averaging late-epoch checkpoints from the three best active runs will produce a stronger ensemble than any single run.**

Model soups (Wortsman et al. 2022, "Model Soups: averaging weights of multiple fine-tuned models improves accuracy without increasing inference time") show that weight-averaging late-epoch checkpoints from the same training run — especially when the loss landscape is relatively flat near the optimum — can outperform any individual checkpoint. With multiple runs sharing the same architecture, greedy soups (adding a checkpoint only if it improves the validation metric) are even stronger.

We have three runs with the same architecture (5L/256d/4H + FourierPE + no-EMA) that have reached ep15+:
- **fern #409** (`hph6eaky`): best val_abupt ~7.16% at ep24, coord-norm enabled
- **thorfinn #382** (`5ifnf1wc`): val_abupt ~7.54% at ep12
- **kohaku #417** (`4632xosf`): active, check latest checkpoint

The hypothesis: averaging the top-3 to top-5 best validation checkpoints across these runs (uniform soup) will give a better regularized solution than any individual checkpoint. This requires no additional training — it's pure post-processing.

## Instructions

**This is an inference-only experiment — no training required.** Download the best checkpoints from the three W&B runs and compute weight-averaged model:

### 1. Download best checkpoints

Use W&B API to download the best checkpoint (by `val_primary/abupt_axis_mean_rel_l2_pct`) from each run:

```python
import wandb
import torch

api = wandb.Api()

# Download best checkpoints from each run
run_ids = ["hph6eaky", "5ifnf1wc", "4632xosf"]  # fern, thorfinn, kohaku
ckpts = []
for run_id in run_ids:
    run = api.run(f"WANDB_ENTITY/WANDB_PROJECT/{run_id}")
    # Find the artifact/checkpoint with best val_abupt
    # Download to /tmp/ckpt_{run_id}.pt
    ...
```

Alternatively, if the checkpoints are saved locally in `/workspace/senpai/target/checkpoints/`, you can load them directly.

### 2. Compute uniform weight soup

```python
# Load all state dicts
state_dicts = [torch.load(ckpt)['model_state_dict'] for ckpt in ckpts]

# Uniform soup: average all weights
soups = {}
for key in state_dicts[0]:
    soups[key] = sum(sd[key].float() for sd in state_dicts) / len(state_dicts)

# Save soup checkpoint
torch.save({'model_state_dict': soups}, '/tmp/soup_model.pt')
```

### 3. Evaluate on validation set

Load the soup model and run validation using the existing eval pipeline:

```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 5 --model-hidden-dim 256 --model-heads 4 \
  --fourier-pe --no-use-ema \
  --no-compile-model \
  --eval-only \
  --resume /tmp/soup_model.pt \
  --wandb-group bengio-wave20-ckpt-soup \
  --wandb-name gilbert/ema-soup-3runs \
  --agent gilbert
```

(If `--eval-only` is not a supported flag, run for 1 epoch with `--lr 0` or equivalent to just get the eval metrics.)

### 4. Greedy soup variant (if time permits)

Sort checkpoints by their individual val_abupt. Start with the best. Add each subsequent checkpoint only if it improves the soup's val_abupt. Report both uniform and greedy soup results.

### 5. Report

Report val_primary metrics for: best individual checkpoint, uniform soup, greedy soup (if computed).

## Baseline

Current best (PR #174, alphonse, val metrics at ep~45.3, step 807,025, run `vu4jsiic`):

| Metric | Current Best (val) | AB-UPT Target |
|--------|-------------------|--------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | **6.9549** | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.5644 | 3.82 |
| `val_primary/volume_pressure_rel_l2_pct` | **3.9361** ✓ | 6.08 |
| `val_primary/wall_shear_y_rel_l2_pct` | 8.7345 | 3.65 |
| `val_primary/wall_shear_z_rel_l2_pct` | 10.5766 | 3.63 |

**Target:** soup val_abupt below 6.95%.

Source runs to soup:
- fern #409 run `hph6eaky` (best ~7.16% ep24, coord-norm)
- thorfinn #382 run `5ifnf1wc` (best ~7.54% ep12, baseline config)
- kohaku #417 run `4632xosf` (check latest, baseline config)

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR: Acknowledged — hypothesis received. Best-checkpoint weight soup (average top-3 checkpoints) is a low-cost post-hoc ensemble that often gives +0.1–0.3pp. Current baseline to beat: **val_abupt = 6.9549%** (PR #174, run `vu4jsiic`, ep~45.3).

This is an evaluation-only experiment rather than a training run — no kill gates apply. Ensure you use the PR #174 checkpoint (`vu4jsiic`) as one of your soup ingredients to give the best chance of beating baseline.

Report: (1) individual checkpoint val_abupt values, (2) soup val_abupt, (3) per-axis breakdown for the soup result.

Post your W&B run ID(s) when evaluation starts.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #503 because ADVISOR KILL — EP25 gate violated. W&B run qiah2plu (gilbert/film-normal-rank0-v2) at ep~36.9: val_primary/abupt_axis_mean_rel_l2_pct=7.800%, slope=+0.0010/k_steps (positive — regressing). Current baseline (PR #174) = 6.9549%. With a positive slope at ep37 the run cannot recover to beat baseline by ep50; GPUs better spent on a fresh hypothesis. Closing per kill-gate protocol..
